pilot/pkg/secrets/kube/secrets.go

```diff
@@ -47,14 +56,139 @@ const (
 
 type SecretsController struct {
 	secrets informersv1.SecretInformer
+	sar     authorizationv1client.SubjectAccessReviewInterface
+
+	clusterID    string
+	remoteGetter RemoteKubeClientGetter
+
+	mu                   sync.RWMutex
+	authorizationCache   map[authorizationKey]authorizationResponse
+	authorizationEnabled bool
+}
+
+type authorizationKey struct {
+	cluster string
+	user    string
+}
+
+type authorizationResponse struct {
+	expiration time.Time
+	authorized error
 }
 
 var _ secrets.Controller = &SecretsController{}
 
-func NewSecretsController(informer informersv1.SecretInformer) *SecretsController {
+type RemoteKubeClientGetter func(clusterID string) kubernetes.Interface
+
+func NewSecretsController(client kube.Client, clusterID string, remoteClientGetter RemoteKubeClientGetter) *SecretsController {
 	// Informer is lazy loaded, load it now
-	_ = informer.Informer()
-	return &SecretsController{informer}
+	_ = client.KubeInformer().Core().V1().Secrets().Informer()
+
+	return &SecretsController{
+		secrets: client.KubeInformer().Core().V1().Secrets(),
+
+		authorizationEnabled: true,
+		sar:                  client.AuthorizationV1().SubjectAccessReviews(),
+		clusterID:            clusterID,
+		remoteGetter:         remoteClientGetter,
+		authorizationCache:   make(map[authorizationKey]authorizationResponse),
+	}
+}
+
+func toUser(serviceAccount, namespace string) string {
+	return fmt.Sprintf("system:serviceaccount:%s:%s", namespace, serviceAccount)
+}
+
+const cacheTTL = time.Minute
+
+// cachedAuthorization checks the authorization cache
+// nolint
+func (s *SecretsController) cachedAuthorization(user, clusterID string) (error, bool) {
+	key := authorizationKey{cluster: clusterID, user: user}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !s.authorizationEnabled {
+		return nil, true
+	}
+	got, f := s.authorizationCache[key]
```

Comment:
potential mem leak,  need to cleanup expired key periodically

Reply:
Each entry is a couple bytes, and we only get an entry on a unique service account. It seems unlikely this will ever grow beyond 5 entries. Do you think its still worth the complexity of periodic cleanup?

Reply:
I guess we can do it on request instead of periodic

Reply:
But the request could not come again when the application deleted. Or it can reconect to another istiod.
One way u can cleanup when client disconnects.

Reply:
The cache is per service account through. Generally you have a very small number of ingress service accounts. I'll add cleanup though

Reply:
hmm, you mean this is only for gateway not sidecars

Reply:
Oh actually in the future we will allow sidecars as well - but probably we will have a different auth method. If we do allow sidecars and this method it will be more important

Reply:
I added eviction